src/main/flight/gps_rescue.c

```diff
@@ -304,19 +319,40 @@ static void rescueAttainPosition()
     /**
         Altitude controller
     */
-    const int16_t altitudeError = (rescueState.intent.targetAltitudeCm - rescueState.sensor.currentAltitudeCm) / 100; // Error in meters
-    const int16_t altitudeDerivative = altitudeError - previousAltitudeError;
+    const int16_t altitudeError = rescueState.intent.targetAltitudeCm - rescueState.sensor.currentAltitudeCm;
+    
+    // P component
+    if (ABS(altitudeError) > 0 && ABS(altitudeError) < GPS_RESCUE_ZVELOCITY_THRESHOLD){
+        scalingRate = (float)altitudeError / GPS_RESCUE_ZVELOCITY_THRESHOLD;
+    } else {
+        scalingRate = 1;
+    }
 
-    // Only allow integral windup within +-15m absolute altitude error
-    if (ABS(altitudeError) < 25) {
-        altitudeIntegral = constrain(altitudeIntegral + altitudeError, -250, 250);
+    if (altitudeError > 0) {
+        zVelocityError = gpsRescueConfig()->ascendRate * scalingRate - rescueState.sensor.zVelocityAvg;
+    } else if (altitudeError < 0) {
+        //decrease vertical speed when the quad is lower than 3mt
+        if (rescueState.sensor.currentAltitudeCm <= 300) {
+            zVelocityError = -GPS_RESCUE_LANDING_DIVE_RATE * scalingRate - rescueState.sensor.zVelocityAvg;
+        } else {
+            zVelocityError = -gpsRescueConfig()->descendRate * scalingRate - rescueState.sensor.zVelocityAvg;
+        }
     } else {
-        altitudeIntegral = 0;
+        zVelocityError = 0;
     }
 
-    previousAltitudeError = altitudeError;
+    // I component
+    if (ABS(zVelocityError) < 100) {
+        zVelocityIntegral = constrain(zVelocityIntegral + zVelocityError / 100, -250, 250);
```

Comment:
Looks like an integer math problem here as `zVelocityError` is int16. With the above limit of `< 100` the I accumulation will always be 0.

Also would be nice to replace some of the constants with defines rather than have "magic numbers" in the code.

Reply:
Also seems like `zVelocityIntegral` should be a float.